### PR TITLE
Update scikit-learn autolog

### PR DESCRIPTION
**Pull request by harupy**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/harupy/mlflow/pull/13236?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/13236/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13236
```

</p>
</details>

### Related Issues/PRs

 #xxx

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: examples/mlflow-3/sklearn_autolog.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# ruff: noqa
+"""
+python examples/sklearn_autolog.py
```
Comment (harupy) [inline examples/mlflow-3/sklearn_autolog.py:3]:
```suggestion
examples/mlflow-3/sklearn_autolog.py
```

File: examples/mlflow-3/sklearn_autolog.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+"""
+
+import logging
```
Comment (dbczumar) [inline examples/mlflow-3/sklearn_autolog.py:6]: @harupy I ran the following example code. Generally seemed to work, but the metrics are missing from the model.

```
# ruff: noqa
"""
python examples/sklearn_autolog.py
"""

import logging
import os
import tempfile

from sklearn.datasets import load_iris
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
import mlflow

os.environ["MLFLOW_AUTOLOGGING_TESTING"] = "true"

mlflow.sklearn.autolog()

X, y = load_iris(return_X_y=True)
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

model = LogisticRegression()
with mlflow.start_run():
    model.fit(X_train, y_train)

outputs = mlflow.get_run(mlflow.last_active_run().info.run_id).outputs
print(outputs)

model_id = outputs.model_outputs[0].model_id
print(model_id)

print(mlflow.get_logged_model(model_id))
```

```
<RunOutputs: model_outputs=[<ModelOutput: model_id='beb58294-3181-4380-b5b3-77f21e4975d6', step=0>]>
beb58294-3181-4380-b5b3-77f21e4975d6
<LoggedModel: artifact_location='file:///Users/corey.zumar/mlflow/mlruns/0/models/beb58294-3181-4380-b5b3-77f21e4975d6/artifacts', creation_timestamp=1727241286667, experiment_id='0', last_updated_timestamp=1727241288677, metrics=[], model_id='beb58294-3181-4380-b5b3-77f21e4975d6', model_type=None, name='model', params={'C': '1.0',
 'class_weight': 'None',
 'dual': 'False',
 'fit_intercept': 'True',
 'intercept_scaling': '1',
 'l1_ratio': 'None',
 'max_iter': '100',
 'multi_class': 'auto',
 'n_jobs': 'None',
 'penalty': 'l2',
 'random_state': 'None',
 'solver': 'lbfgs',
 'tol': '0.0001',
 'verbose': '0',
 'warm_start': 'False'}, run_id='df30a6d2091a4ba2be7bdf536d1552b0', status='ModelStatus.READY', status_message=None, tags={}>
```

File: examples/mlflow-3/sklearn_autolog.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+# ruff: noqa
```
Comment (dbczumar) [inline examples/mlflow-3/sklearn_autolog.py:1]: @harupy Can you add a similar example that uses GridSearch and make sure that the outputs make sense (e.g. correct loggedmodels are produced, linked to the correct runs, and have the correct metrics?)

Comment (harupy) [reply examples/mlflow-3/sklearn_autolog.py:1]: Added!

File: examples/mlflow-3/sklearn_autolog.py (hunk 2 above)
Comment (harupy) [reply examples/mlflow-3/sklearn_autolog.py]: https://github.com/mlflow/mlflow/pull/13236/commits/2636db78bdaedac19ada1249378f56c0d1bec44e fixed this issue

File: mlflow/sklearn/__init__.py [hunk 4]
```diff
@@ -1576,20 +1559,44 @@ def _log_model_with_except_handling(*args, **kwargs):
...
+            pos_label=pos_label,
+            dataset=dataset,
+            model_id=model_id,
+        )
+        if y is None and not logged_metrics:
+            _logger.warning(
+                "Training metrics will not be recorded because training labels were not specified."
+                " To automatically record training metrics, provide training labels as inputs to"
+                " the model training function."
+            )
```
Comment (harupy) [inline mlflow/sklearn/__init__.py:1593]: moved this block after model creation to use the model ID.

**Pull request merged**